js/src/edit-free-campaign/index.js

```diff
@@ -178,6 +178,40 @@ export default function EditFreeCampaign() {
 		'/google/dashboard'
 	);
 
+	const handleChooseAudienceChange = ( change, newTargetAudience ) => {
+		updateTargetAudience( newTargetAudience );
+
+		/**
+		 * If users removed a target audience country,
+		 * or changed the target audience location from 'all' to 'selected',
+		 * we remove the country from shipping rates and shipping times.
+		 */
+		if (
+			newTargetAudience.countries.length <
+				targetAudience.countries.length ||
```

Comment:
Thanks for catching that! 👍 

Now I'm thinking, instead of putting the code under `handleChooseAudienceChange`, we can put it inside `handleChooseAudienceContinue` (when users click on the "Continue" button) and `handleStepClick` (when users go from Step 1 to Step 2). This means that if users remove a country and add them back and then leave, the unsaved data prompt should not appear. I'll work on the change.

Reply:
@eason9487 , turns out the issue in your video happens in the `develop` branch too. The problem is that even though in the UI the form fields look the same and there are "no changes" to the users, the underlying data `targetAudience.countries` and `savedTargetAudience.countries` have the same country codes but in different orders. 

https://user-images.githubusercontent.com/417342/163433622-3900e98f-a045-4332-90ab-b9a60e2286e0.mp4

What do you think? 🤔

Reply:
> turns out the issue in your video happens in the `develop` branch too.

Thanks for the finding! I didn't notice that. 😂 

> Now I'm thinking, instead of putting the code under `handleChooseAudienceChange`, we can put it inside `handleChooseAudienceContinue` (when users click on the "Continue" button) and `handleStepClick` (when users go from Step 1 to Step 2).

> I'm thinking we can still proceed with the changes I proposed above, and open another issue for dealing with the problem in the video separately. Sounds good?

It sounds good to open another issue for the incorrect comparison of `targetAudience.countries`.

However, I'm not so sure would it be a good idea to move the code to `handleChooseAudienceContinue` and `handleStepClick`. The shipping rates and times could be restored by using `savedShippingRates` and `savedShippingTimes` in the `handleChooseAudienceChange` instead. And, with the new comparison function `getDifferentShippingTimes` added in #1420, the `didTimesChanged` could be changed to be compared in the same way as shipping rates. In this way, it could more appropriately determine if there is unsaved shipping data.

Reply:
> The shipping rates and times could be restored by using `savedShippingRates` and `savedShippingTimes` in the `handleChooseAudienceChange` instead.

Interesting. So I guess you are saying when we remove a country and add the removed one back, we try to find the shipping rate and shipping time for that country from `savedShippingRates` and `savedShippingTimes` and restore it back, right?

I thought of one scenario which might end up being funny / buggy:

1. Let's say we have a country "JP" with shipping rate 12 and shipping time 12. (These values are stored in `savedShippingRates` and `savedShippingTimes`.)
2. We go to Step 2, change the values to shipping rate 10 and shipping time 10. (These values are stored in the "form" / `useState`.)
3. We go to Step 1, remove the country JP. (Here, form useState would no longer have the shipping rate 10 and shipping time 10 for JP, because we remove them in `handleChooseAudienceChange`.)
4. We add the country JP back in.
    - Here, we should restore back the shipping rate and shipping time value for JP. What should they be: 12 or 10? 🤔 I'm thinking it should be the latest value 10, but we already lose that value here.
 	
> And, with the new comparison function `getDifferentShippingTimes` added in #1420, the `didTimesChanged` could be changed to be compared in the same way as shipping rates. In this way, it could more appropriately determine if there is unsaved shipping data.

Yup, that makes sense. I'll look into that. 🙂 👍

Reply:
> I thought of one scenario which might end up being funny / buggy: [...]

That's a very detailed consideration! 👍  I didn't think of this scenario. Maybe moving the filtering of shipping rates/times to the render layer could handle all the form operation combinations. For example:

```js
const finalCountries = getFinalCountries( targetAudience );
const [ formShippingRates, formShippingTimes ] = useMemo( () => {
	const countries = finalCountries || [];
	return [
		shippingRates.filter( ( { country } ) =>
			countries.includes( country )
		),
		shippingTimes.filter( ( { countryCode } ) =>
			countries.includes( countryCode )
		),
	];
}, [ finalCountries, shippingRates, shippingTimes ] );

// ... omitted

const loadedShippingRates = ! hfrShippingRates ? null : formShippingRates;
const loadedShippingTimes = ! hfrShippingTimes ? null : formShippingTimes;

// ... omitted

const didRatesChanged = hasUnsavedShippingRates(
	formShippingRates,
	savedShippingRates
);
```

Reply:
> Maybe moving the filtering of shipping rates/times to the render layer could handle all the form operation combinations. For example: [...]

I think that is going to make things more complicated. 😅 With that approach, we will have the following states:

- `savedShippingRates`: data from backend API.
- `shippingRates`: this will then act like a "cached" / "staging" area to store shipping rates. When users modify shipping rates values in Step 2, this will be updated.
- `formShippingRates`: we get this by filtering `shippingRates` against selected countries in Step 1. 

Then, here's another scenario which might end up looking buggy:

1. Go to Edit Free Listings.
2. In Step 1, users add a new country "SG".
3. In Step 2, users add shipping rate 22 and shipping time 22 for "SG". (`shippingRates` is updated here.)
4. Go back to Step 1, and remove country "SG". 
5. Go to Step 2, "SG" is not visible in shipping rates and shipping times. Users would think it has been removed. (However, "SG" is still in `shippingRates`. It is not visible in UI because we are using `formShippingRate`.)
6. Go back to Step 1, add back the country "SG".
7. Go to Step 2, users would see shipping rate 22 and shipping time 22 for "SG", which looks like a bug.

---

After some thinking from the above scenarios, I think an important thing to point out is about _the switching between Step 1 and Step 2_. When users switch from one step to another, it acts like a confirmation that "I am done with this step, you can apply the changes and show me the updated things in the next step". What this means is that I think it makes sense to put the state update inside `handleChooseAudienceContinue` (when users click on the "Continue" button) and `handleStepClick` (when users go from Step 1 to Step 2) (as per my earlier comment https://github.com/woocommerce/google-listings-and-ads/pull/1409#discussion_r850596474).

I'll give this a try and we see how it goes. 😄

Reply:
> It sounds good to open another issue for the incorrect comparison of `targetAudience.countries`.

👍 I've created issue https://github.com/woocommerce/google-listings-and-ads/issues/1431 for this.

Reply:
> I'll give this a try and we see how it goes. 😄

I made the changes in [`cf9a72f` (#1409)](https://github.com/woocommerce/google-listings-and-ads/pull/1409/commits/cf9a72ffec704e9abf2509358fb64227ceb5d22b), by doing the state update inside `handleChooseAudienceContinue` (when users click on the "Continue" button) and `handleStepClick` (when users go from Step 1 to Step 2).

Reply:
> I think that is going to make things more complicated.

If the requirement asked for handling that scenario, then the complexity from the code perspective was not a blocker for implementation.

> Then, here's another scenario which might end up looking buggy:
> 
> [...]
> 6. Go back to Step 1, add back the country "SG".
> 7. Go to Step 2, users would see shipping rate 22 and shipping time 22 for "SG", which looks like a bug.

If the form value is entered by the user, then it could also be interpreted as convenient handling, which let the user don't need to re-enter the value again when they try to add back a country.

> After some thinking from the above scenarios, I think an important thing to point out is about _the switching between Step 1 and Step 2_. When users switch from one step to another, it acts like a confirmation that "I am done with this step, you can apply the changes and show me the updated things in the next step".

Thanks for the explanation! This looks reasonable as well.